tuning.py

```diff
@@ -0,0 +1,317 @@
+                   single_mut_offset: float = 0.0) -> float:
+    """Returns the epistatic offset tuning parameter to achieve `fraction_reciprocal_adaptive_epistasis`.
+
+    # TODO(nthomas) add note about pretuning the landscape
+    """
+    # single mutant offset changes the set of adaptive singles
+    pretuned_landscape = potts_model.PottsModel(landscape.weight_matrix,
+                                                landscape.field_vec,
+                                                landscape.wildtype_sequence,
+                                                single_mut_offset=single_mut_offset)
+    doubles_df = get_doubles_df(pretuned_landscape, threshold=0.0, adaptive=True)
+    # we want fraction to remain negative
+    epi_offset = -1 * np.quantile(doubles_df.residual, q=fraction_reciprocal_adaptive_epistasis)
+    # np.quantile returns float64 by default
+    return epi_offset.astype(np.float32)
+
+
+def get_coupling_scale(landscape: potts_model.PottsModel,
+                       epistatic_horizon: float,
+                       field_scale: float = 1.0,
+                       single_mut_offset: float = 0.0,
+                       epi_offset: float = 0.0) -> float:
+    """Returns the coupling scale tuning parameter to achieve `epistatic_horizon`.
+
+    The epistatic horizon depends on other landscape statistics.
+    Let $s_+$ be the average adaptive single mutant effect.
+    Let $e_{+,+}$ be the average epistatic effect for a pair of adaptive singles.
+    Then, given K, we solve the following equation to determine the coupling-scale:
+
+    $$K * field_scale (s_+ + field-scale) + (K choose 2) * coupling-scale (e_{+,+} + epi_offset) = 0$$
+    """
+    adaptive_threshold = 0.0
+    is_adaptive = True
+
+    pretuned_landscape = potts_model.PottsModel(landscape.weight_matrix,
+                                                landscape.field_vec,
+                                                wt_seq=landscape.wildtype_sequence,
+                                                field_scale=field_scale,
+                                                single_mut_offset=single_mut_offset,
+                                                epi_offset=epi_offset)
+    mean_adaptive_single_effect = get_mean_single_effect(
+        pretuned_landscape, threshold=adaptive_threshold, adaptive=is_adaptive)
+    mean_epistasis_effect = get_doubles_df(
+        pretuned_landscape, threshold=adaptive_threshold, adaptive=is_adaptive).residual.mean()
+
+    K = epistatic_horizon
+    numerator = - K * mean_adaptive_single_effect
+    denominator = comb(K, 2) * mean_epistasis_effect
+    coupling_scale = numerator / denominator
+    return coupling_scale
+
+
+# TODO(nthomas) implement fraction_reciprocal_deleterious_epistasis:
+#   The fraction of deleterious(-, -) doubles that exhibit positive epistasis.
+def get_tuning_kwargs(landscape: potts_model.PottsModel,
+                      fraction_adaptive_singles: Optional[float] = None,
```

Comment:
I found it confusing that these are all optional. What happens when a value of None is passed?

Reply:
I want to be able to tune some aspects of the landscape without tuning others. So when None is passed, that statistic is left untuned. I'll update the docstring to reflect this.